machineconfiguration/v1/types.go

```diff
@@ -760,3 +760,77 @@ type ContainerRuntimeConfigList struct {
 
 	Items []ContainerRuntimeConfig `json:"items"`
 }
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// PinnedImageSet describes a set of images that should be pinned.
+//
+// Compatibility level 1: Stable within a major release for a minimum of 12 months or 3 minor releases (whichever is longer).
+// +openshift:compatibility-gen:level=1
+type PinnedImageSet struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// +required
+	Spec PinnedImageSetSpec `json:"spec"`
+	// +optional
+	Status PinnedImageSetStatus `json:"status"`
+}
+
+// PinnedImageSetSpec defines the desired state of a PinnedImageSet.
+type PinnedImageSetSpec struct {
+	// MachineConfigPoolSelector selects which pools the PinnedImageSet should apply to.
+	// A nil selector will result in no pools being selected.
+	// +optional
+	MachineConfigPoolSelector *metav1.LabelSelector `json:"machineConfigPoolSelector,omitempty"`
```

Comment:
It can be more than 2 as custom pools are a feature[1] but I would imagine most clusters operator with only 2.

It would be possible to label the pools IE and thus define the elector to match `pinned-image-set-opt-in: true`
```
metadata:
  labels:
    pinned-image-set-opt-in: true
```

[1] PinnedImageSet describes a set of images that should be pinned.

Reply:
We could explicitly define an enum something like.
```
 {Worker,Master}PinnedImageSetMachineConfigPool
 ```
 I would like to enforce one set per pool so this would work from my perspective.

Reply:
If there are custom pools, then I don't think an enum will work, as that will prevent users from working with the custom pools.

Enforcing a "one set per pool" will be very difficult if you have a selector, with multiple entries in the list, how would you enforce that the selectors do not overlap? And, even if they don't overlap at admission time, what's to stop the labels being modified later and then having them overlap outside of the admission control

Could this be a property on the machine config pool itself?

Or perhaps, a resource that's `.metadata.name` must match a machine pool name?

Having the `.metadata.name` match ensures your uniqueness that you desire

The alternative is to have some sort of maybe policy and binding, in theory they could be two separate CRDs, or they could be in a cluster wide singleton somewhere

```
pinnedImages:
  policies:
  - name: generic
    pinnedImages:
    - ....
    - ....
  bindings:
  - poolName: worker
    policyName: generic
```

Reply:
Another suggestion from David, keep the image lists as CRs, and then reference the image list from the MCP, solves having multiple lists, and being able to bind specifically one of those lists to the MCP

Reply:
makes sense good feedback